Add explanation, update example, minor fixes (#761)
cosmos/sdk-tutorials: interchain-exchange/tutorial/00-tutorial.md

Tokens can be bought or sold with Limit Orders on a simple order book, there is no notion of Liquidity Pool or AMM.

The market is unidirectional: the token sold on the source chain cannot be bought back as it is, and the token bought from the target chain cannot be sold back using the same pair. If a token on a source chain is sold, it can only be bought back by creating a new pair on the order book. This is due to the nature of IBC, creating a `voucher` token on the target blockchain. In this tutorial you will learn the difference of a native blockchain token and a `voucher` token that is minted on another blockchain. You will learn how to create a second order book pair in order to receive the native token back.

For example: There are two `ibc` enabled chains - Earth_chain and Venus_chain. 

In the next chapter you will learn details about the design of the interblockchain exchange.

# App Design

This section describes how the interchain exchange module is designed. The module has order books, buy orders, and sell orders. First, an order book for a pair of tokens must be created. After an order book exists, you can create buy and sell orders for this pair of tokens.

> The module will make use of the Inter-Blockchain Communication Standard [IBC](https://ibcprotocol.org/). With the use of IBC, the module can create order books for tokens to have multiple blockchains interact and exchange their tokens.

Reviewer: ```suggestion
The module will make use of the Inter-Blockchain Communication protocol [IBC](https://ibcprotocol.org/). With the use of IBC, the module can create order books for tokens to have multiple blockchains interact and exchange their tokens.
```

- Send IBC packets from one blockchain to another
- Deal with timeouts and acknowledgements of IBC packets

## Concepts good to know

**Order Books**

> **Order books** are electronic list of *buy* and *sell* orders for specific financial instrument - For eg: cryptocurrencies.
> An order book lists the number of tokens being *bid* (buy) on or *offered* (sell) at each price point, or market depth. It also identifies the market participants behind the buy and sell orders. These lists help traders and also improve market transparency because they provide valuable trading information.

Reviewer 2: ```suggestion
**Order books** are electronic list of *buy* and *sell* orders for specific financial instrument - For eg: cryptocurrencies. An order book lists the number of tokens being *bid* (buy) on or *offered* (sell) at each price point, or market depth. It also identifies the market participants behind the buy and sell orders. These lists help traders and also improve market transparency because they provide valuable trading information.
```

## How the module works

You will learn how to build an exchange that works with two or more blockchains. The module is called `ibcdex`.

The module allows to open an exchange order book between a pair of token within a blockchain network and between a pair of token on another blockchain network. The blockchains are required to have the `ibcdex` module available.

Tokens can be bought or sold with Limit Orders on a simple order book, there is no notion of Liquidity Pool or AMM.

The market is unidirectional: the token sold on the source chain cannot be bought back as it is, and the token bought from the target chain cannot be sold back using the same pair. If a token on a source chain is sold, it can only be bought back by creating a new pair on the order book. This is due to the nature of IBC, creating a `voucher` token on the target blockchain. In this tutorial you will learn the difference of a native blockchain token and a `voucher` token that is minted on another blockchain. You will learn how to create a second order book pair in order to receive the native token back.

> <!-- The above comment on line 27 is a bit hard to follow. Adding an example below for easier understanding.-->

Reviewer 2: ```suggestion
```

Comments are visible to users, let's not add them, unless they are meant for end-users.

> For example: There are two `ibc` enabled chains - Earth_chain and Venus_chain. 

Reviewer 2: ```suggestion
For example: There are two IBC-enabled chains - Earth chain and Venus chain. 
```

Earth_chain is confusing.

In the next chapter you will learn details about the design of the interblockchain exchange.

# App Design

In this chapter you will learn how the interchain exchange module is designed. The module has **order books**, **buy** and **sell orders.** First, an order book for a pair of token has to be created. After an order book exists, you can create buy and sell orders for this pair of token.

> The module will make use of the Inter-Blockchain Communication Standard [IBC](https://github.com/cosmos/ics/blob/master/ibc/2_IBC_ARCHITECTURE.md). With use of the IBC, the module can create order books for tokens to have multiple blockchains interact and exchange their tokens. <!-- The link provided gives a 404. -->

Reviewer 2: ```suggestion
The module will make use of the Inter-Blockchain Communication Standard [IBC](https://ibcprotocol.org/). With the use of IBC, the module can create order books for tokens to have multiple blockchains interact and exchange their tokens.
```

## Performing an Exchange with a Sell Order

> <!-- This is a bit difficult to follow. Using a simpler example to illustrate exchange with sell order. Simplifying the calculations by using coherent values for a token and integer amount -->

Reviewer: before merging, let's delete all comments in the Markdown files because these comments are visible to all since our repo is open source, thanks @alijnmerchant21

The counterparty (sender of the buy mcx order) received 5 mcx vouchers. vcx balance hasn't changed, because the correct amount of vcx (50) were locked at the creation of the buy order during the previous step.

```yml
# Target blockchain
balances:
- amount: "5" # increased from 0
  denom: ibc/99678A10AF684E33E88959727F2455AE42CCC64CD76ECFA9691E1B5A32342D33 # mcx voucher
```

> -->

Reviewer: ```suggestion
```

Reviewer: we probably want to delete this end comment tag?

```
# Sell order broadcasted to the source chain
interchanged tx ibcdex send-sellOrder ibcdex channel-0 mcx 20 vcx 5
```

The sell order (for 5 `marscoin` at 5 `venuscoin`) was filled on the target chain by the buy order. The amount of the buy order on the target chain has decreased by 5 `marscoin`.

```yml
# Target blockchain
BuyOrderBook:
- amountDenom: mcx
  creator: ""
  index: ibcdex-channel-0-mcx-vcx
  orderIDTrack: 5
  orders:
  - amount: 15 # decreased from 20
    creator: cosmos1qlrz3peenc6s3xjv9k97e8ef72nk3qn3a0xax2
    id: 3
    price: 5
  priceDenom: vcx
```

The sender of the filled sell order exchanged 5 `marscoin` for 25 `venuscoin` vouchers. 25 vouchers is a product of the amount of the sell order (5 `marscoin`) and price of the buy order (5 `venuscoin`).

```yml
# Source blockchain
balances:
- amount: "25" # increased from 0
  denom: ibc/50D70B7748FB8AA69F09114EC9E5615C39E07381FE80E628A1AF63A6F5C79833 # vcx voucher
- amount: "985" # decreased from 990
  denom: mcx
```

> The counterparty (sender of the buy mcx order) received 5 `marscoin` vouchers. `venuscoin` balance hasn't changed, because the correct amount of `venuscoin` (50) were locked at the creation of the buy order during the previous step.

Reviewer: do we mean mcx here? Or did we decide to spell out marscoin?

Author: I suggest to keep `marscoin` and `venuscoin` when explaining. For code we can use mcx and vcx

Reviewer: oooh I know it's work, but I'm in agreement with Denis on this... is there a way to use consistent token names and denom? Teach me!

The counterparty (sender of the buy mcx order) received 5 mcx vouchers. vcx balance hasn't changed, because the correct amount of vcx (50) were locked at the creation of the buy order during the previous step.

```yml
# Target blockchain
balances:
- amount: "5" # increased from 0
  denom: ibc/99678A10AF684E33E88959727F2455AE42CCC64CD76ECFA9691E1B5A32342D33 # mcx voucher
```
-->

> We now have two orders open for `marscoin`: a *sell order* on the **source chain** (for 20 `marscoin` at 5 `venuscoin`) and a *buy order* on the **target chain** (for 5 `marscoin` at 5 `venuscoin`). Let's perform an exchange by sending a sell order to the source chain.

Reviewer 2: If we switch to `marscoin`, we need to change the denom in config as well as in examples.

Author: We can still use mcx and vcx while writing code. But for explanation I propose on using `marscoin` and `venuscoin` respectively.

Reviewer 2: Arguably, this can confuse people, because we use `code` syntax, so either 1. `marscoin` everywhere, or 2. `mcx` everywhere, 3. or `mcx` and Mars coin.

Reviewer: I vote for same in code and text, and yes, using the tick marks for code syntax for literals

> In this chapter you will implement the logic for creating order books.

Reviewer: ```suggestion
In this section, you implement the logic for creating order books.
```

> In Cosmos SDK the state is stored in a key-value store. Each order book will be stored under a unique key composed of four values: port ID, channel ID, source denom and target denom. For example, an order book for `mcx` (`marscoin`) and `vcx` (`venuscoin`) could be stored under `ibcdex-channel-4-mcx-vcx`. Define a function that returns an order book store key.

Reviewer: ```suggestion
In Cosmos SDK, the state is stored in a key-value store. Each order book is stored under a unique key that is composed of four values: port ID, channel ID, source denom, and target denom. 

For example, an order book for `mcx` (`marscoin`) and `vcx` (`venuscoin`) could be stored under `ibcdex-channel-4-mcx-vcx`. Define a function that returns an order book store key.
```

Reviewer: let's focus on present tense, even with tutorials, I know we struggle with this in all of our tutorials, see https://developers.google.com/style/tense
